### ISSUE TITLE
bug in angleBetween()

### ISSUE THREAD

**Issue by agnius-vasiliauskas** [processing/p5.js#3973]
angleBetween() returns angle between `[0, PI]`, but should return `[-PI,PI]`.
Because now a **pair** of directions can have same angle, however they are different:
Attaching explanation / picture how should be and why.
![angles](https://user-images.githubusercontent.com/4015780/63761275-86610e00-c8c9-11e9-9b0f-e1fae9ccad06.png)

**Comment by lmccart** [comment]
@0x69 can you please provide a code snippet that illustrates the issue? thanks

**Comment by 0x69** [comment]
Of course:

```
function setup() {
 let v0 = createVector(1, 0);
 let v1 = createVector(0, 1);
 let v2 = createVector(0, -1);  
 let a1 = v0.angleBetween(v1);
 let a2 = v0.angleBetween(v2);
 print(degrees(a1) + '° between v0,v1');
 print(degrees(a2) + '° between v0,v2');
}
```

>  BAD output:
>  90° between v0,v1 
>  90° between v0,v2
>  EXPECTED output :
>  a)
>  -90° between v0,v1 
>  90° between v0,v2
>  OR
>  b)
>  90° between v0,v1 
>  -90° between v0,v2
>  Depends on how API will choose negative angle tactics - when going
>  clockwise or anti-clockwise from v0 vector to target vector
>

**Comment by nishantkr18** [comment]
@0x69 it seems that `angleBetween()` always returns the absolute angle, although one can know the difference pointed by you, using `let a1 = v0.angleBetween(v2)*Math.sign(v0.cross(v2).z);` when dealing with only x and y (two dimensions).

**Comment by CrypticGuy** [comment]
Hi @stalgiag 
I would like to work on this.

I think it can be solved by either providing the method mentioned by @nishantkr18 in the reference/docs, or we could add a parametrized option to the function, where the angle is displayed clockwise or counter-clockwise, depending on the argument passed. 
If none of the above methods then a new function could be defined to get the counter or counter-clockwise angle.
If these seem like suitable methods, I would like to work on this issue.

**Comment by 0x69** [comment]
> @0x69 it seems that `angleBetween()` always returns the absolute angle, although one can know the difference pointed by you, using `let a1 = v0.angleBetween(v2)*Math.sign(v0.cross(v2).z);` when dealing with only x and y (two dimensions).

Nice solution ! However there are some edge cases. In case both vectors are parallel to each other - angles _0_ and _π_, then cross product results in **zero length vector**, thus rendering bug,- showing 0 angle in both edge cases. Fix to this problem we can consider zero as positive sign :

`let fullAngle = v1.angleBetween(v2) * Math.sign(v1.cross(v2).z || 1);`

**Comment by singhvisha** [comment]
Hi @stalgiag ,Can I work on this issue?

**Comment by stalgiag** [comment]
Hi @singhvisha, I think that @CrypticGuy might have asked first.

@CrypticGuy I think that an optional boolean parameter should work well.

**Comment by 0x69** [comment]
@stalgiag 
>I think that an optional boolean parameter should work well.

Depends on what this parameter will do. If it will change only sign of displayed angle from positive to negative as @CrypticGuy suggested ("_angle is displayed clockwise or counter-clockwise, depending on the argument passed_") - it's pointless, because in such case in effect it is the same as showing only absolute angle (because sign will not change automatically) !
The only reasonable option in function would be: "absolute_angle (true/false)"

- absolute_angle = true => result is displayed in old fashion way [0,π]
- absolute_angle = false =>  result is displayed in range [-π, +π] (as @nishantkr18 suggested and with my fix too)

**Comment by CrypticGuy** [comment]
Yes, I agree this makes more sense.

**Comment by stalgiag** [comment]
@0x69 Ah got it. Thanks for the explanation. A boolean argument doesn't make sense to me if the purpose is so that these are equivalent:
`vec1.angleBetween(vec2, true)`
&&
`abs(vec1.angleBetween(vec2)`

Processing also returns a value between `[0, PI]` which is probably why the function was designed this way initially. I am open to modifying the function so that it returns a value between `[-PI, PI]`. 
Do you see any possible negative consequences of returning the full `[-PI, PI]` range? My instinct is to say no because it is much easier to get from a value in that range back to an absolute value but just want to check since I haven't been able to find any information on why that design decision was made initially and I haven't used this function extensively in my own sketches.

**Comment by 0x69** [comment]
@stalgiag

Indeed, not much point in this additional parameter. Function just could return [-PI,PI] instead without any additional parameter added. The ones who will want absolute angle will call `abs()` like you said. However, there is a small reason of keeping this new parameter idea still. Suppose, somebody has written angle check `if (angle == PI)`, now with modification when function returns [-PI,PI] this user script will be broken. Then she/he will need a fix `if (abs(angle) == PI)`. I.e. our change will be not backwards compatible, we may break existing user code. However solution with default parameter:
`vec1.angleBetween(vec2, absoluteAngle = true)` will not break existing user code. It's the only reason in having additional optional parameter which I can think of.

**Comment by CrypticGuy** [comment]
#4048

**Comment by 0x69** [comment]
I just saw comment in code
> // For 3D vectors there is no sense of direction as such

Need to add that 3D vector direction perfectly makes sense too ! Just vector rotation in 3D space is defined not by a scalar angle, but by a vector composed of two components - angles α, β which are rotations of 3D vector **projections** in two perpendicular planes. See example :
![bitmap jpg](https://user-images.githubusercontent.com/4015780/65964818-54157400-e466-11e9-877c-fb056bc06325.png)
So technically function `vec3d1.angleBetween(vec3d2)` _could_ work, just one issue that maybe we would need to rename it "rotationBetween()" huhu ? Another issue that it should compute both α, β angles and return result as 2D vector `[α, β]`.

**Comment by stalgiag** [comment]
closed via #4048 

@agnius-vasiliauskas sounds cool! Feel free to open an issue for a `rotationBetween()` function for further discussion.
